### ISSUE TITLE
Semantics of `intros ->` when the left-hand side is not a variable

### ISSUE THREAD

**Issue by charguer** [coq/coq#13520]
I'd like to open a discussion on the semantics of `intros ->`.

The documentation says that this pattern "replaces the right-hand side of the hypothesis with the left-hand side of the hypothesis in the conclusion of the goal, the hypothesis is cleared", the point is that the substitution does not affect the hypotheses in the general case.

However, there is an exception to that rule: variables are given a special treatment: "if the left-hand side of the hypothesis is a variable, it is substituted everywhere in the context and the variable is removed.".

Does anyone know what was the motivation for not making "substitution everywhere in the goal" the behavior in all cases, and not just in the cases of variables? (Was this simply inherited from ssr?)

I believe there is a problem with the fact of clearing an equality that has been substituted only in the goal and not in the context. The problem is that it can turn provable goal into non-provable goals at a moment where you don't expect it---at least, I didn't expect it.

Here is a simple example illustrating the issue. In the first lemma, `intros ->` fails, so we know that we need to go try something else, such as `intros <-`. In the second example, however, `intros ->` does not fail, yet gets you stuck by making an important piece of information vanish.

```
Parameter val : Type.
Parameter err : val.

Lemma test: forall (v:val),
  (v <> err) ->
  (err = v) ->
  True.
Proof using.
  intros v H.
  try intros ->. (* fails cause no occurence of [val_error] in goal, only context *)
  intros <-. (* succeeds because [v1] is a variable, so it is substituted in context *)
Abort.

Lemma vicious_test:  forall (v:val),
  (v <> err) ->
  (err = v) ->
  (err <> err).
Proof using.
  intros v H.
  intros ->. (* succeeds, but results in unprovable goal *)
Abort.
```

I can see two ways out:
- One would be to make `intros ->` not clear the hypothesis that it substitutes in case the left-hand side is not a variable; this may require an extra "clear" in many proof scripts for proper bookkeeping.
- One would be to make `intros ->` to perform substitutions also in the context; obviously, this would break a number of existing proof scripts.

Maybe someone else can think of a better solution to improve the current situation, that is, the unexpected silent disappearence of useful information?

**Comment by herbelin** [comment]
I agree with you, and actually, I thought that I had implemented option 1 (i.e.  `intros ->` not clear the hypothesis that it substitutes in case the left-hand side is not a variable), since, even when rewriting a non-variable everywhere, we can lose in provability. Consider e.g. `Lemma test f g : f 0 = 0 -> g 0 = 0 -> f (g (f 0)) = 0.`. If I chose for clearing, it is probably that this was the most interesting to do in practice (compared to explicitly using a `rewrite` instead).

Regarding « the motivation for not making "substitution everywhere in the goal" », it is maybe a bit by conservatism but also because I was unsure that this was always what we wanted, and probably a bit also because I was unsure that the pattern-matching algorithm was good enough to recognize subterms up to small details such as slightly different implicit arguments or so. I'd be ready to implement the other strategy to see how better it would suit in general (using the CI as feedback).

**Comment by charguer** [comment]
@herbelin Thanks for your feedback. Let's schedule this discussion for the next coqdev meeting. I will be interesting to hear everyone's opinion on what should be the specification, and perhaps most importantly, what is the general guideline that applies here to justify the choice of a particular specification.

**Comment by herbelin** [comment]
We can also use a room at the [CUDW](https://github.com/coq/coq/wiki/Coq-Users-and-Developers-Workshop-2020) if you want.

**Comment by Zimmi48** [comment]
My intuition is that "intro patterns" always "destruct" the variable that they are applied to. Thus, it's perfectly normal and acceptable that `->` will clear the hypothesis. I didn't know that it didn't rewrite everywhere, probably because I never encountered this case, but it would also seem more normal to me that it did.

BTW, I just learned by reading the changelog, that rewrite intro patterns were introduced in 8.1 and last modified in 8.5:
>New introduction patterns "->" and "<-" for immediate rewriting of introduced hypotheses.

>Behavior of introduction patterns -> and <- made more uniform (hypothesis is cleared, rewrite in hypotheses and conclusion and erasing the variable when rewriting a variable).

**Comment by herbelin** [comment]
Slightly diverging, a common interesting case to consider is the case of rewriting expressions of the form `(x,y) = t` which are equivalent to the conjunction of `x = fst t` and `y = snd t` and where there is no other possible intent than rewriting every `x` and `y` (which are not already part of an expression `(x,y)`, so as to save an η-expansion) into `fst t` and `snd t`. Then, there is no reason to keep the variables and hypothesis, which should be cleared.
